src/rulesets/LinearAlgebra/matfun.jl

```diff
@@ -0,0 +1,261 @@
+    P = copy(Inn)
+    W = C[2] * P
+    V = C[1] * P
+    Apows = typeof(P)[]
+    for k in 1:(div(size(C, 1), 2) - 1)
+        k2 = 2 * k
+        P *= A2
+        push!(Apows, P)
+        W += C[k2 + 2] * P
+        V += C[k2 + 1] * P
+    end
+    U = A * W
+    X = V + U
+    F = lu!(V - U)  # NOTE: use lu! instead of LAPACK.gesv! so we can reuse factorization
+    ldiv!(F, X)
+    Xpows = typeof(X)[X]
+    if si > 0  # squaring to reverse dividing by power of 2
+        for t in 1:si
+            X *= X
+            push!(Xpows, X)
+        end
+    end
+
+    # Undo the balancing
+    for j in ilo:ihi
+        scj = scale[j]
+        for i in 1:n
+            X[j, i] *= scj
+        end
+        for i in 1:n
+            X[i, j] /= scj
+        end
+    end
+
+    if ilo > 1  # apply lower permutations in reverse order
+        for j in (ilo - 1):-1:1
+            LinearAlgebra.rcswap!(j, Int(scale[j]), X)
+        end
+    end
+    if ihi < n  # apply upper permutations in forward order
+        for j in (ihi + 1):n
+            LinearAlgebra.rcswap!(j, Int(scale[j]), X)
+        end
+    end
+    return X, (ilo, ihi, scale, C, si, Apows, W, F, Xpows)
+end
+
+# Application of the chain rule to exp!, also Algorithm 6.4 from
+# Al-Mohy, Awad H. and Higham, Nicholas J. (2009).
+# Computing the Fréchet Derivative of the Matrix Exponential, with an application to
+# Condition Number Estimation", SIAM. 30 (4). pp. 1639-1657.
+# http://eprints.maths.manchester.ac.uk/id/eprint/1218
+function _matfun_frechet!(
+    ::typeof(exp), A::StridedMatrix{T}, X, ΔA, (ilo, ihi, scale, C, si, Apows, W, F, Xpows)
+) where {T<:BlasFloat}
+    n = LinearAlgebra.checksquare(A)
+    for j in ilo:ihi
+        scj = scale[j]
+        for i in 1:n
+            ΔA[j, i] /= scj
+        end
+        for i in 1:n
+            ΔA[i, j] *= scj
+        end
+    end
+
+    if si > 0
+        ΔA ./= convert(T, 2^si)
+    end
+
+    ∂A2 = mul!(A * ΔA, ΔA, A, true, true)
+    A2 = first(Apows)
+    # we will repeatedly overwrite ∂temp and ∂P below
+    ∂temp = Matrix{eltype(∂A2)}(undef, n, n)
+    ∂P = copy(∂A2)
+    ∂W = C[4] * ∂P
+    ∂V = C[3] * ∂P
+    for k in 2:(length(Apows) - 1)
+        k2 = 2 * k
+        P = Apows[k - 1]
+        ∂P, ∂temp = mul!(mul!(∂temp, ∂P, A2), P, ∂A2, true, true), ∂P
+        axpy!(C[k2 + 2], ∂P, ∂W)
+        axpy!(C[k2 + 1], ∂P, ∂V)
+    end
+    ∂U, ∂temp = mul!(mul!(∂temp, A, ∂W), ΔA, W, true, true), ∂W
+    ∂temp .= ∂U .- ∂V
+    ∂X = add!!(∂U, ∂V)
+    mul!(∂X, ∂temp, first(Xpows), true, true)
+    ldiv!(F, ∂X)
+
+    if si > 0
+        for t in 1:(length(Xpows) - 1)
+            X = Xpows[t]
+            ∂X, ∂temp = mul!(mul!(∂temp, X, ∂X), ∂X, X, true, true), ∂X
+        end
+    end
+
+    for j in ilo:ihi
+        scj = scale[j]
+        for i in 1:n
+            ∂X[j, i] *= scj
+        end
+        for i in 1:n
+            ∂X[i, j] /= scj
+        end
+    end
+
+    if ilo > 1  # apply lower permutations in reverse order
```

Comment:
I haven't been able to find a matrix `A` that hits this branch. As far as I can tell, all `exp` tests in LinearAlgebra also do not hit the corresponding branch, but code coverage indicates it is covered, so it must be indirectly hit by one of the trigonometric/hyperbolic matrix function tests or something.

Reply:
Nevermind, found one